src/ucp/dt/datatype_iter.inl

```diff
@@ -500,10 +542,29 @@ ucp_datatype_iter_is_end(const ucp_datatype_iter_t *dt_iter)
 static UCS_F_ALWAYS_INLINE void
 ucp_datatype_memh_dereg(ucp_context_h context, ucp_mem_h *memh_p)
 {
-    if (*memh_p != NULL) {
-        ucp_memh_put(context, *memh_p);
-        *memh_p = NULL;
+    if (*memh_p == NULL) {
+        return;
+    }
+
+    /* Check if memh is external (user memh), and if so, don't touch it */
+    if (ucp_memh_is_user_memh(*memh_p)) {
+        return;
     }
+
+    /* If it's internal, release it */
+    ucp_memh_put(context, *memh_p);
+    *memh_p = NULL;
+}
+
+static UCS_F_ALWAYS_INLINE ucs_status_t
+ucp_datatype_iter_check_memh_mds(ucp_context_h context, ucp_mem_h memh,
+                                 ucp_md_map_t md_map, unsigned uct_flags)
+{
+    if (!ucs_test_all_flags(memh->md_map, md_map)) {
+        return ucp_memh_init_uct_reg(context, memh, uct_flags, md_map);
```

Comment:
What happens with `memh->parent` in this case? Can't it be a memory leak?

Reply:
yes there is, so it seems.
@yosefe WDYT?

Reply:
IMO it's doable: can add to parent if it exists, then copy to memh itself